src/test/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/unit/service/ApplicationServiceTest.kt

```diff
@@ -331,4 +336,149 @@ class ApplicationServiceTest {
+      .withId(applicationId)
+      .withCreatedByUser(user)
+      .withSubmittedAt(OffsetDateTime.now())
+      .produce()
+      .apply {
+        schemaUpToDate = true
+      }
+
+    every { mockUserService.getUserForRequest() } returns user
+    every { mockApplicationRepository.findByIdOrNull(applicationId) } returns application
+    every { mockJsonSchemaService.checkSchemaOutdated(application) } returns application
+
+    val result = applicationService.submitApplication(applicationId, "{}", username)
+
+    assertThat(result is AuthorisableActionResult.Success).isTrue
+    result as AuthorisableActionResult.Success
+
+    assertThat(result.entity is ValidatableActionResult.GeneralValidationError).isTrue
+    val validatableActionResult = result.entity as ValidatableActionResult.GeneralValidationError
+
+    assertThat(validatableActionResult.message).isEqualTo("This application has already been submitted")
+  }
+
+  @Test
+  fun `submitApplication returns Success, runs json logic rules and creates assessment`() {
+    val applicationId = UUID.fromString("fa6e97ce-7b9e-473c-883c-83b1c2af773d")
+    val username = "SOMEPERSON"
+
+    val newestSchema = ApprovedPremisesApplicationJsonSchemaEntityFactory().produce()
+
+    val user = UserEntityFactory()
+      .withDeliusUsername(username)
+      .produce()
+
+    val application = ApplicationEntityFactory()
+      .withApplicationSchema(newestSchema)
+      .withId(applicationId)
+      .withCreatedByUser(user)
+      .withSubmittedAt(null)
+      .produce()
+      .apply {
+        schemaUpToDate = true
+      }
+
+    every { mockUserService.getUserForRequest() } returns user
+    every { mockApplicationRepository.findByIdOrNull(applicationId) } returns application
+    every { mockJsonSchemaService.checkSchemaOutdated(application) } returns application
+    every { mockJsonSchemaService.validate(newestSchema, application.data!!) } returns true
+    every { mockApplicationRepository.save(any()) } answers { it.invocation.args[0] as ApplicationEntity }
+    every { mockJsonLogicService.resolveBoolean(newestSchema.isPipeJsonLogicRule, application.data!!) } returns true
+    every { mockJsonLogicService.resolveBoolean(newestSchema.isWomensJsonLogicRule, application.data!!) } returns false
+    every { mockAssessmentService.createAssessment(application) } returns AssessmentEntityFactory()
+      .withApplication(application)
+      .withAllocatedToUser(user)
+      .produce()
+
+    val result = applicationService.submitApplication(applicationId, "{}", username)
+
+    assertThat(result is AuthorisableActionResult.Success).isTrue
+    result as AuthorisableActionResult.Success
+
+    assertThat(result.entity is ValidatableActionResult.Success).isTrue
+    val validatableActionResult = result.entity as ValidatableActionResult.Success
+    assertThat(validatableActionResult.entity.isPipeApplication).isTrue
+    assertThat(validatableActionResult.entity.isWomensApplication).isFalse
```

Comment:
I don't quite follow how these resolve to `True` and `False` here - might be better if we're specific when setting up the schema

Reply:
the return values are hardcoded on lines 464/465 since `JsonLogicService` is responsible for correctly resolving the answers.  i.e. in this test the json logic rules on the schema that's produced on line 443 don't really have any effect so figured it'd perhaps be misleading to set them up to look realistic there.

We are covering the full flow in the integration test too though (i.e. the rules are actually read from the schema and applied against the application's data)

Reply:
Ah, yeah, right you are. That's fine then 👍  Knew I was missing _something_!